### PR TITLE
Router controller

### PR DESCRIPTION
**Pull request by Aoxian**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-[![Build Status](https://travis-ci.org/Aoxian/HTTPServer.svg?branch=master)](https://travis-ci.org/Aoxian/HTTPServer/branches)
+[![Build Status](https://travis-ci.org/Aoxian/HTTPServer.svg?branch=router-controller)](https://travis-ci.org/Aoxian/HTTPServer/branches)
```
Comment (nystromb) [inline README.md]: Are you doing this just so the build status shows up in the readme on the branch? I dont think this is necessary considering we see this information on github checks.

File: src/Aoxian.Web.HTTP/Server.cs [hunk 2]
```diff
@@ -5,15 +5,17 @@ namespace Aoxian.Web.HTTP
...
-        private static Handler _handler;
-        private string _host;
-        private int _port;
+        private readonly string _host;
+        private readonly int _port;
+        private static IRouter _router;
+        private static IRequestParser _requestParser;
 
-        public Server(string host, int port, Handler handler)
+        public Server(string host, int port, IRouter router, IRequestParser requestParser)
```
Comment (nystromb) [inline src/Aoxian.Web.HTTP/Server.cs]: I like that you're passing the request parser in here. But something to consider **if** you keep finding yourself adding parameters, is that you might want to consider creating a single "ServerConfiguration" object that has functions to return all the server configuration stuff. Not saying you should do this right now, but it may make sense in the future. 4 params is becoming a lot of things to pass in.

File: src/Aoxian.Web.HTTP/Router.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        {
+            _routes = new List<Route>();
+        }
+
+        public IResponse Route(IRequest request)
+        {
+            try
+            {
+                return _routes.Find(route => request.Path == route.Path)
+                    .Controller.Handle(request);
```
Comment (nystromb) [inline src/Aoxian.Web.HTTP/Router.cs]: 🔥

File: src/Aoxian.Web.HTTP/RequestParser.cs [hunk 4]
```diff
@@ -31,10 +27,10 @@ public Request Parse(byte[] requestBuffer)
             };
         }
 
-        private Dictionary<string, string> ParseHeaders(string rawRequest)
+        private static Dictionary<string, string> ParseHeaders(string rawRequest)
```
Comment (nystromb) [inline src/Aoxian.Web.HTTP/RequestParser.cs]: Why did you add the `static` modifier to these functions?

File: README.md (hunk 1 above)
Comment (Aoxian) [reply README.md]: You got it.  I don't know a better way to do this.  I could leave it all `master`, but then if something is failing its a false positive.

File: src/Aoxian.Web.HTTP/RequestParser.cs (hunk 4 above)
Comment (Aoxian) [reply src/Aoxian.Web.HTTP/RequestParser.cs]: Hmm. I've run up against this while pairing with Tyler.  As I refactor things, the IDE I'm using (Rider) will suggest changes and sometimes automatically implements them for me. In this case, I do not remember explicitly thinking "I need to make this static" and implementing the change.  I did intentionally rename the constants in this file though.

File: src/Aoxian.Web.HTTP/Router.cs (hunk 3 above)
Comment (Aoxian) [reply src/Aoxian.Web.HTTP/Router.cs]: 😄

File: README.md (hunk 1 above)
Comment (nystromb) [reply README.md]: If I were you, I wouldn't worry about it. Dont make yourself do extra work :)

File: src/Aoxian.Web.HTTP/RequestParser.cs (hunk 4 above)
Comment (nystromb) [reply src/Aoxian.Web.HTTP/RequestParser.cs]: To me, I think it's really odd. Because `private` means you can't call the function outside of the class. But `public` means that you can call this function publicly on the class. They kind of have the opposite meaning. If theres a public method that is static, then these will have to be `static` also.

Unless there's a reason to keep them, I'd remove the modifier.

File: src/Aoxian.Web.HTTP/Router.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            }
+            catch
+            {
+                return new Response();
```
Comment (nystromb) [inline src/Aoxian.Web.HTTP/Router.cs]: I see you're defaulting to 404 here. I think it might be better to be explicit here that you're returning 404, instead of having to go look in the `Response` to see what exactly is going on.

Comment (Aoxian) [reply src/Aoxian.Web.HTTP/Router.cs]: Good catch; fixing this now.

File: src/Aoxian.Web.HTTP/RequestParser.cs (hunk 4 above)
Comment (Aoxian) [reply src/Aoxian.Web.HTTP/RequestParser.cs]: I can remove them. I think Rider just wants as many things to be as static as possible; that said, I am not trying to rely on the IDE as much as skill.
